graph/converter.py

```diff
@@ -0,0 +1,357 @@
+        y_row = y + 1
+        x_row = x
+        for node in self.tree.children:
+            preprocessed_graph_data['nodes'].append(
+                converter(node)._create_node(x_row, y_row))
+            preprocessed_graph_data['edges'].append(converter(node)._create_edge(
+                self.tree.node_id, node.node_id, preprocessed_graph_data['nodes'][-1]))
+            x_row = x_row + 1
+            if node.children is not None:
+                preprocessed_graph_data = converter(node)._help_to_sigma_dict(
+                    x_row + 1, y_row + 1, preprocessed_graph_data)
+        return self._fix_graph(preprocessed_graph_data)
+
+    def _count_max_y(self, out):
+        max_y = 0
+
+        for node in out['nodes']:
+            if max_y < node['y']:
+                max_y = node['y']
+        return max_y
+
+    def _create_nodes_in_rows(self, rows):
+        nodes_in_rows = dict()
+
+        for i in range(rows + 1):
+            nodes_in_rows[i] = []
+        return nodes_in_rows
+
+    def _push_nodes_to_nodes_in_row(self, out, nodes_in_rows):
+        for node in out['nodes']:
+            nodes_in_rows[node['y']].append(node)
+
+    def _remove_empty_rows(self, nodes_in_rows, max_y):
+        for row in range(max_y + 1):
+            if not nodes_in_rows[row]:
+                del nodes_in_rows[row]
+
+    def _move_rows(self, nodes_in_rows):
+        count = 0
+        nodes_in_rows1 = dict()
+
+        for row in nodes_in_rows:
+            nodes_in_rows1[count] = nodes_in_rows[row]
+            for node in nodes_in_rows1[count]:
+                node['y'] = count
+            count += 1
+        return nodes_in_rows1
+
+    def _create_positions(self, nodes_in_rows):
+        positions = []
+        for row in nodes_in_rows:
+            len_of_row = len(nodes_in_rows[row])
+            if len_of_row > 1:
+                if (len_of_row % 2) == 1:
+                    len_of_row += 1
+
+                for i in range((int(-(len_of_row / 2))) * 2,
+                               (int(+(len_of_row / 2)) + 1) * 2, 2):
+                    positions.append(i)
+
+                if len_of_row == 2:
+                    positions.remove(0)
+
+                if len(nodes_in_rows[row]) < len(positions):
+                    positions.pop()
+                    if len(nodes_in_rows[row]) < len(positions):
+                        positions.pop(0)
+
+                count = 0
+
+                for pos in positions:
+                    nodes_in_rows[row][count]['x'] = pos
+                    count += 1
+                positions = []
+            else:
+                nodes_in_rows[row][0]['x'] = 0
+
+        return positions
+
+    def _convert_nodes_in_rows_to_nodes(self, nodes_in_rows):
+        nodes = []
+        for row in nodes_in_rows:
+            for node in nodes_in_rows[row]:
+                nodes.append(node)
+        return nodes
+
+    def _change_position(self, nodes_in_rows):
```

Comment:
Code in this function is quite complex. It may not be worth to change, but let's keep an eye on code like this in the future.